compiler/rustc_monomorphize/src/collector.rs

```diff
@@ -329,6 +307,42 @@ impl<'tcx> UsageMap<'tcx> {
     }
 }
 
+struct MonoItems<'tcx> {
+    // We want a set of MonoItem + Span where trying to re-insert a MonoItem with a different Span
+    // is ignored. Map does that, but it looks wrong.
+    items: FxIndexMap<MonoItem<'tcx>, Span>,
+}
+
+impl<'tcx> MonoItems<'tcx> {
+    fn new() -> Self {
+        Self { items: FxIndexMap::default() }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.items.is_empty()
+    }
+
+    fn push(&mut self, item: Spanned<MonoItem<'tcx>>) {
+        // Insert only if the entry does not exist. A normal insert would stomp the first span that
+        // got inserted.
+        self.items.entry(item.node).or_insert(item.span);
+    }
+
+    fn extend(&mut self, iter: impl IntoIterator<Item = Spanned<MonoItem<'tcx>>>) {
+        for item in iter {
+            self.push(item);
+        }
+    }
+
+    fn into_iter(self) -> impl Iterator<Item = Spanned<MonoItem<'tcx>>> {
```

Comment:
Can you just implement the normal `IntoIterator` trait here? The calls to `into_iter` in the for loops below seem awkward.